interactive_engine/executor/ir/core/src/catalogue/canonical_label.rs

```diff
@@ -0,0 +1,582 @@
+            for i in 0..vertex_group.len() {
+                let current_v_id: PatternId = vertex_group[i];
+                for j in (i + 1)..vertex_group.len() {
+                    match self.cmp_vertices(pattern, current_v_id, vertex_group[j])? {
+                        Ordering::Greater => vertex_group_tmp_vec[i] += 1,
+                        Ordering::Less => vertex_group_tmp_vec[j] += 1,
+                        Ordering::Equal => (),
+                    }
+                }
+
+                let v_group: PatternId = vertex_group_tmp_vec[i];
+                if v_group != initial_group {
+                    has_converged = false;
+                }
+
+                updated_vertex_group_map.insert(current_v_id, v_group);
+                updated_vertex_groups
+                    .entry((v_label, v_group))
+                    .and_modify(|vertex_group| vertex_group.push(current_v_id))
+                    .or_insert(vec![current_v_id]);
+            }
+        }
+
+        // Update vertex group manager
+        self.vertex_group_map = updated_vertex_group_map;
+        self.vertex_groups = updated_vertex_groups;
+        self.has_converged = has_converged;
+
+        // Update the order of vertex adjacencies
+        self.update_vertex_adjacencies_order();
+        Ok(())
+    }
+}
+
+impl CanonicalLabelManager {
+    /// Set unique ranks to each vertex and edge
+    pub fn pattern_ranking(&mut self, pattern: &mut Pattern) -> IrPatternResult<()> {
+        let start_v_id: PatternId = self.get_pattern_ranking_start_vertex(pattern)?;
+        let mut next_free_vertex_rank: PatternId = 0;
+        let mut next_free_edge_rank: PatternId = 0;
+        self.vertex_rank_map
+            .insert(start_v_id, Some(next_free_vertex_rank));
+        next_free_vertex_rank += 1;
+        let mut visited_edges: BTreeSet<PatternId> = BTreeSet::new();
+        // Initialize Stack for adjacencies
+        let mut adjacency_stack: VecDeque<Adjacency> =
+            self.init_adjacencies_stack(start_v_id, &self.vertex_adjacencies_map)?;
+        // Perform DFS on adjacencies
+        while let Some(adjacency) = adjacency_stack.pop_back() {
+            // Insert edge to dfs sequence if it has not been visited
+            let adj_edge_id: PatternId = adjacency.get_edge_id();
+            if visited_edges.contains(&adj_edge_id) {
+                continue;
+            }
+            visited_edges.insert(adj_edge_id);
+            self.edge_rank_map
+                .insert(adj_edge_id, Some(next_free_edge_rank));
+            next_free_edge_rank += 1;
+
+            // Set dfs id to the vertex if it has not been set before
+            let current_v_id: PatternId = adjacency.get_adj_vertex().get_id();
+            let is_vertex_visited = self
+                .vertex_rank_map
+                .get(&current_v_id)
+                .unwrap_or(&None)
+                .is_some();
+            if !is_vertex_visited {
```

Comment:
`if veretx_rank_map.containsKey(&current_v_id) { ... }`
do not need `is_vertex_visited` above